src/main/java/dev/slimevr/filtering/QuaternionMovingAverage.java

```diff
@@ -0,0 +1,84 @@
+package dev.slimevr.filtering;
+
+import com.jme3.math.FastMath;
+import com.jme3.math.Quaternion;
+
+
+public class QuaternionMovingAverage {
+
+	private final float smoothFactor;
+	private final float predictFactor;
+	private final CircularArrayList<Quaternion> rotBuffer;
+	private final Quaternion quatBuf = new Quaternion();
+	private final Quaternion targetQuat = new Quaternion();
+	private final Quaternion lastQuaternion;
+	private final Quaternion filteredQuaternion;
+	private static final float SMOOTH_QUADRATIC_DIVIDER = 11f;
+	private static final float SMOOTH_QUADRATIC_MIN = 0.022f;
+	private static final float PREDICT_QUADRATIC_DIVIDER = 3.1f;
+	private static final float PREDICT_QUADRATIC_MIN = 0.11f;
+	private static final int PREDICT_BUFFER = 6;
+
+
+	public QuaternionMovingAverage(
+		TrackerFilters type,
+		float amount,
+		Quaternion initialRotation
+	) {
+		amount = Math.max(amount, 0);
+
+		if (type == TrackerFilters.SMOOTHING) {
+			smoothFactor = FastMath.pow(1 - amount, 2) / SMOOTH_QUADRATIC_DIVIDER
+				+ SMOOTH_QUADRATIC_MIN;
+		} else {
+			// smooths a little to reduce jitter
+			smoothFactor = 0.08f - (amount / 90f);
+		}
+
+		if (type == TrackerFilters.PREDICTION) {
+			rotBuffer = new CircularArrayList<>(PREDICT_BUFFER);
+			predictFactor = FastMath.pow(amount, 2) / PREDICT_QUADRATIC_DIVIDER
+				+ PREDICT_QUADRATIC_MIN;
+		} else {
+			rotBuffer = null;
+			predictFactor = 0;
+		}
+
+		filteredQuaternion = new Quaternion(initialRotation);
+		lastQuaternion = new Quaternion(initialRotation);
+	}
+
+	// 1000hz
+	synchronized public void update() {
+		targetQuat.set(lastQuaternion);
+
+		// Prediction
+		if (rotBuffer != null) {
+			if (rotBuffer.size() > 0) {
+				quatBuf.set(targetQuat);
+				rotBuffer.forEach(quatBuf::multLocal);
+				targetQuat.slerpLocal(quatBuf, predictFactor);
+			}
+		}
+
+		// Smoothing
+		filteredQuaternion.slerpLocal(targetQuat, smoothFactor);
```

Comment:
Just updating on this comment, it still holds true with the latest changes, and I think the first paragraph is right in saying it's not great, but the alternative would have it just freezing when there's no next position as opposed to its current behaviour of slowing down... I don't think there's any solution other than just using what I talked about in the second paragraph that would make it interpolate smoothly handling inconsistent timing.

So unless you can write some linear smoothing code that just switches to prediction when the next frame is late, then I approve of the current method with hope that there can be a secondary solution later.

Reply:
The slowing down and never reaching its destination isn’t a problem. However, at high smoothing and low/unstable TPS, the speeding/slowing down issue could be noticed indeed.
I would like to look into doing it linearly like you’re saying, but I might do that in another PR in the future…

Reply:
Here's a visual if it helps at all, if the speeding/slowing is intended, then it's implemented just fine and I'll leave it as approved, I would love to see something like this in the future though
![Smoothing](https://user-images.githubusercontent.com/5095026/191169715-26b3dd4d-4db3-4cf5-86f7-1cbd2bcf464d.jpg)

Reply:
I think this might not be as complicated as I thought to implement now that I see it.
Just need to get the average FPS (and average TPS too maybe) with the NanoTimer and then do very simple math off that.
I’ll try it when I have time :3

Reply:
Done, please resolve if what I did looks good to you :3 
https://github.com/SlimeVR/SlimeVR-Server/pull/241/commits/f4ee059a1c4764b8ec4f8367ddbccda48f641f34